### PR TITLE
Add token renewal per request

### PR DESCRIPTION
**Pull request by janakamarasena**
### Proposed changes in this pull request

Improvments for https://github.com/wso2/product-is/issues/4253

This improvement provides the ability to renew the access token and the refresh token(where applicable) per each token request and revoke previously available active token for a matching clientid, user and scopes combination.

To enable token renewal per request add the following configuration inside the **/<OAuth>** tag in the identity.xml file located in <is_home>/repository/conf/identity

  `<RenewTokenPerRequest>true</RenewTokenPerRequest>`

 Note that this will not affect the flow when self-contained access tokens are used(which by default renews access tokens per request) or when using the refresh token grant type(which renews the access token by default and the refresh token depending on the 'RenewRefreshTokenForRefreshGrant' configuration located in the identity.xml file).

### Follow up actions
Document the feature

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java [hunk 1]
```diff
@@ -127,6 +132,22 @@ public OAuth2AccessTokenRespDTO issue(OAuthTokenReqMessageContext tokReqMsgCtx)
             }
             // Return a new access token in each request when JWTTokenIssuer is used.
             if (accessTokenNotRenewedPerRequest(oauthTokenIssuer, tokReqMsgCtx)) {
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java]: We might need to document these scenarios properly. By looking at this code, it seems like there are multiple cases where new access token can be issued. To me accessTokenNotRenewedPerRequest and isTokenRenewalPerRequestEnabled are confusing.

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java]: accessTokenNotRenewedPerRequest is used when JWTTokenIssuer is used and only renews the access token. isTokenRenewalPerRequestEnabled will actually revoke the existing access token and refresh token so a new token pair will be generated.

Comment (hasinthaindrajee) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java]: accessTokenNotRenewedPerRequest is not just JWTTokenIssuer specific it seems. It's something which is coming from token issuer. Do we have these documented in a developer guide or somewhere ?

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java]: Mentioned JWTTokenIssuer only in the previous comment as only that class has overridden the renewAccessTokenPerRequest method from the interface and sends "true". For others the default value false is sent. If someone uses the [OAuthTokenGenerator extension point ](https://docs.wso2.com/display/IS570/Extension+Points+for+OAuth#ExtensionPointsforOAuth-OAuthTokenGenerator) and overrides the renewAccessTokenPerRequest methods to send "true" then isTokenRenewalPerRequestEnabled will not be used.  Not sure on the documentation, will check on how we can document this clearly.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 2]
```diff
@@ -2657,5 +2659,23 @@ private static IdentityProvider getResidentIdp(String tenantDomain) throws Ident
...
 
+    /**
+     * Used to build an OAuth revocation request DTO.
+     *
+     * @param oAuthClientAuthnContext  OAuth client authentication context.
+     * @param accessToken Access token to be revoked.
+     * @return Returns a OAuth revocation request DTO.
+     */
+    public static OAuthRevocationRequestDTO buildOAuthRevocationRequest(OAuthClientAuthnContext oAuthClientAuthnContext,
+                                                                        String accessToken){
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: formatting

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1128

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 3]
```diff
@@ -2596,6 +2611,18 @@ private void parseUseSPTenantDomainConfig(OMElement oauthElem) {
         }
     }
 
+    private void parseTokenRenewalPerRequestConfiguration(OMElement oauthConfigElem) {
```
Comment (Kanapriya) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Add method level comment

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java [hunk 4]
```diff
@@ -871,6 +894,48 @@ private static void removeTokenFromCache(OAuthCacheKey cacheKey, AccessTokenDO t
             }
         }
     }
+
+    private static void revokeExistingToken(String clientId, String accessToken) throws IdentityOAuth2Exception {
```
Comment (Kanapriya) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java]: Add method level comment

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 3 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1128

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java (hunk 4 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/util/ResponseTypeHandlerUtil.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1128

**Pull request merged**